### PR TITLE
feat: Support OpenAI reasoning effort configuration for O1/O3 models

### PR DESCRIPTION
**Pull request by da-moon**
# Reason for This Change

OpenAI's O1 and O3 models support configuring reasoning effort to control the
model's behavior. This change implements and improves support for reasoning
effort configuration through model name suffixes, making it easier to adjust
model behavior without modifying code.

### `crates/goose/src/providers/formats/openai.rs`

1. Core Functionality:

   - Added reasoning effort extraction from model name suffixes (e.g.,
     o3-mini-high → high)
   - Implemented default "medium" reasoning effort when not specified
   - Ensured proper model name handling for both O-family and non-O-family
     models
   - Added proper validation for o1-mini model (blocked as unsupported)

2. Model Name Processing:

   - Improved model name handling with separate logic for O-family models
   - Added robust parsing of model name suffixes
   - Maintained original model names for non-O family models
   - Implemented clean separation of model name and reasoning effort

3. Documentation and Code Structure:

   - Added comprehensive module-level documentation explaining all features
   - Added detailed function-level documentation with examples
   - Split large functions into smaller, well-documented components
   - Improved code organization with clear sections
   - Added type annotations for better error handling clarity

4. Test Coverage:

   - Added TestModelConfig struct for easier test case creation
   - Added assert_request helper for consistent test validation
   - Split tests into smaller, focused test functions
   - Added comprehensive test cases for:
     - O1 and O3 model configurations
     - Default and custom reasoning efforts
     - Non-O family model handling
     - Edge cases and error conditions
   - Improved test naming and organization
   - Added specific tests for model name handling
   - Added float comparison with appropriate epsilon values

5. Request Generation:

   - Updated token field usage for O family models
   - Removed temperature support for O1/O3 models
   - Maintained backward compatibility for non-O family models
   - Improved error messages for unsupported configurations

6. Code Quality Improvements:

   - Extracted common test setup into helper functions
   - Added clear separation between test categories
   - Improved error message clarity
   - Enhanced code readability with better variable names
   - Added proper documentation for all public functions
   - Implemented proper type annotations throughout

7. Internal Refactoring:

   - Split the message formatting logic into smaller functions
   - Added helper functions for processing different content types
   - Improved error handling with proper type annotations
   - Enhanced code organization with logical grouping

These changes provide a robust and well-documented implementation of the
reasoning effort feature while improving the overall code quality and
maintainability of the OpenAI format module.

## Validation Criteria

Test Run

```console
~/tmp/goose> cargo test --package goose --lib providers::formats::openai -- --nocapture
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.19s
     Running unittests src/lib.rs (target/debug/deps/goose-411d7bfe04e9d139)

running 24 tests
test providers::formats::openai::tests::test_create_request_o1_reasoning_effort ... ok
test providers::formats::openai::tests::test_create_request_o3_reasoning_effort ... ok
test providers::formats::openai::tests::test_format_tools ... ok
test providers::formats::openai::tests::test_format_messages ... ok
test providers::formats::openai::tests::test_format_tools_empty ... ok
test providers::formats::openai::tests::test_gpt35_turbo_config ... ok
test providers::formats::openai::tests::test_format_tools_duplicate ... ok
test providers::formats::openai::tests::test_gpt4_standard_config ... ok
test providers::formats::openai::tests::test_gpt4_with_version_suffix ... ok
test providers::formats::openai::tests::test_non_o_family_with_high_suffix ... ok
test providers::formats::openai::tests::test_non_o_family_with_low_suffix ... ok
test providers::formats::openai::tests::test_o1_default_reasoning_effort ... ok
test providers::formats::openai::tests::test_o1_mini_not_supported ... ok
test providers::formats::openai::tests::test_o1_custom_reasoning_effort ... ok
test providers::formats::openai::tests::test_non_o_family_with_medium_suffix ... ok
test providers::formats::openai::tests::test_o3_custom_reasoning_effort ... ok
test providers::formats::openai::tests::test_response_to_message_text ... ok
test providers::formats::openai::tests::test_o3_default_reasoning_effort ... ok
test providers::formats::openai::tests::test_o3_invalid_suffix_defaults_to_medium ... ok
test providers::formats::openai::tests::test_response_to_message_invalid_func_name ... ok
test providers::formats::openai::tests::test_response_to_message_json_decode_error ... ok
test providers::formats::openai::tests::test_response_to_message_valid_toolrequest ... ok
test providers::formats::openai::tests::test_format_messages_multiple_content ... ok
test providers::formats::openai::tests::test_format_messages_complex ... ok

test result: ok. 24 passed; 0 failed; 0 ignored; 0 measured; 65 filtered out; finished in 0.01s
```

### CODE REVIEW

**Comment by michaelneale** [comment]
oh this is really cool - the openai formats are used by a few providers so we want to be careful with that., but I like this.

**Comment by michaelneale** [comment]
@da-moon looks like need some `cargo fmt` but some compile errors too - if you push again I think it will trigger CI

**Comment by da-moon** [comment]
I am pretty sure compilation worked as I actually used all 3 modes of O3-mini and ran a bunch of end to end tests and goose sessions. I might have made some breaking changes at one point, that's why the build is failing. Let me look into it and see why it is failing. 

Sadly, the models were not as fine-tuned on tool calls as I expected them to and Sonnet was still better.

**Comment by da-moon** [comment]
I will work on handling formatting and dealing with the build issue tomorrow morning.

@michaelneale what are your thoughts on a pr that enables option to use prompt caching or turn it off?

**Comment by da-moon** [comment]
@michaelneale what are your thoughts on creating a docker image for building the binary so that there is a unified build environment in both CI and local machine? I can open a different PR for that and submit it by EOD tomorrow

**Comment by da-moon** [comment]
@michaelneale The compilation issue was due to failing doctest. it should be fixed now

- Library Unit Tests 

```console
~/tmp/goose> cargo test --package goose --lib providers::formats::openai -- --nocapture
   Compiling goose v1.0.4 (/home/damoon/tmp/goose/crates/goose)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 11.60s
     Running unittests src/lib.rs (target/debug/deps/goose-411d7bfe04e9d139)

running 24 tests
test providers::formats::openai::tests::test_format_tools ... ok
test providers::formats::openai::tests::test_gpt35_turbo_config ... ok
test providers::formats::openai::tests::test_format_messages ... ok
test providers::formats::openai::tests::test_format_tools_empty ... ok
test providers::formats::openai::tests::test_gpt4_standard_config ... ok
test providers::formats::openai::tests::test_non_o_family_with_low_suffix ... ok
test providers::formats::openai::tests::test_create_request_o1_reasoning_effort ... ok
test providers::formats::openai::tests::test_non_o_family_with_medium_suffix ... ok
test providers::formats::openai::tests::test_o1_custom_reasoning_effort ... ok
test providers::formats::openai::tests::test_create_request_o3_reasoning_effort ... ok
test providers::formats::openai::tests::test_format_tools_duplicate ... ok
test providers::formats::openai::tests::test_o1_default_reasoning_effort ... ok
test providers::formats::openai::tests::test_non_o_family_with_high_suffix ... ok
test providers::formats::openai::tests::test_o3_custom_reasoning_effort ... ok
test providers::formats::openai::tests::test_o1_mini_not_supported ... ok
test providers::formats::openai::tests::test_gpt4_with_version_suffix ... ok
test providers::formats::openai::tests::test_o3_default_reasoning_effort ... ok
test providers::formats::openai::tests::test_response_to_message_text ... ok
test providers::formats::openai::tests::test_o3_invalid_suffix_defaults_to_medium ... ok
test providers::formats::openai::tests::test_response_to_message_json_decode_error ... ok
test providers::formats::openai::tests::test_response_to_message_invalid_func_name ... ok
test providers::formats::openai::tests::test_response_to_message_valid_toolrequest ... ok
test providers::formats::openai::tests::test_format_messages_multiple_content ... ok
test providers::formats::openai::tests::test_format_messages_complex ... ok
```

- Doc Tests

```console
~/tmp/goose> cargo test --doc -- --test-threads=1
   Compiling serial_test v3.2.0
   Compiling ciborium v0.2.2
   Compiling goose v1.0.4 (/home/damoon/tmp/goose/crates/goose)
   Compiling criterion v0.5.1
    Finished `test` profile [unoptimized + debuginfo] target(s) in 3.37s
   Doc-tests goose

running 10 tests
test crates/goose/src/config/base.rs - config::base::Config (line 68) - compile ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::create_request (line 561) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::format_messages (line 277) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::format_text_content (line 40) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::format_tool_request (line 71) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::format_tool_response (line 190) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::format_tools (line 352) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::get_usage (line 503) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::process_tool_content (line 134) ... ok
test crates/goose/src/providers/formats/openai.rs - providers::formats::openai::response_to_message (line 413) ... ok

test result: ok. 10 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 43.00s

   Doc-tests goose_mcp

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests mcp_client

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests mcp_core

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests mcp_macros

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests mcp_server

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

**Comment by michaelneale** [comment]
@da-moon I like the idea of prompt caching, has worked well elsewhere we have used it so yeah - looiks like needs another fmt)

File: crates/goose/src/providers/formats/openai.rs [hunk 1]
```diff
@@ -249,24 +542,82 @@ pub fn get_usage(data: &Value) -> Result<Usage, ProviderError> {
...
 
-    let is_o1 = model_config.model_name.starts_with("o1");
-    let is_o3 = model_config.model_name.starts_with("o3");
-
     let system_message = json!({
-        "role": if is_o1 { "developer" } else { "system" },
+        // NOTE: per OPENAI docs , With O1 and newer models, `developer`
+        // should replace `system` role .
+        // https://platform.openai.com/docs/api-reference/chat/create
+        "role": if is_o1 || is_o3{ "developer" } else { "system" },
```
Comment (salman1993) [inline crates/goose/src/providers/formats/openai.rs:620]: nit: add whitespace after is_o3

File: crates/goose/src/providers/formats/openai.rs [hunk 2]
```diff
@@ -128,18 +326,53 @@ pub fn format_messages(messages: &[Message], image_format: &ImageFormat) -> Vec<
...
+/// This function formats tools according to the OpenAI API tool specification.
+/// It handles tool names, descriptions, and parameters, ensuring they are properly
+/// formatted for the OpenAI API.
+///
+/// # Arguments
+/// * `tools` - The tools to format
+///
+/// # Returns
+/// * `anyhow::Result<Vec<Value>, Error>` - A vector of formatted tools ready for the OpenAI API,
+/// or an error if there are duplicate tool names
```
Comment (salman1993) [inline crates/goose/src/providers/formats/openai.rs]:
```suggestion
///   or an error if there are duplicate tool names
```

^ this is why build is failing on`clippy` warning

File: crates/goose/src/providers/formats/openai.rs (hunk 1 above)
Comment (salman1993) [inline crates/goose/src/providers/formats/openai.rs]:
```suggestion
        "role": if is_o1 || is_o3 { "developer" } else { "system" },
```

nit: adding whitespace

**Review by salman1993 — COMMENTED** [review]
LGTM! Thanks a lot for adding this in. 🙏🏼  Left 2 suggestions which should make the CI checks pass

**Comment by jasonkneen** [comment]
Already did this, tested and worked but it was rejected because I used AI to do it (irony right?)

**Comment by salman1993** [comment]
@da-moon i have incorporated your changes in this PR - https://github.com/block/goose/pull/1266
